lightly/api/api_workflow_upload_metadata.py

```diff
@@ -172,22 +178,34 @@ def upload_custom_metadata(self,
             raise ValueError("There exist image names in the custom metadata "
                              "without corresponding filenames on the server.")
 
-        if verbose:
-            # wrap samples in a progress bar
-            samples = tqdm.tqdm(samples)
-
-        for sample in samples:
-
-            metadata = filename_to_metadata[sample.file_name]
-
-            if metadata is not None:
-                # create a request to update the custom metadata of the sample
+        # retry upload if it times out
+        def upload_sample_metadata(request, sample):
+            return retry(
+                self._samples_api.update_sample_by_id,
+                request,
+                dataset_id=self.dataset_id,
+                sample_id=sample.id,
+            )
+
+        # limit number of concurrent requests
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = []
```

Comment:
What is the reason you used this futures implementation of parallel upload? It is unknown for me and thus hard to review.
I think it would be better to use the same workflow as for uploading the samples: 
https://github.com/lightly-ai/lightly/blob/4daa75625082a36511e45a3bf7f82c9512eea36b/lightly/api/api_workflow_upload_dataset.py#L196-L198

Reply:
Oh, I just named it `futures` instead of `results` but it is the same thing :)

Reply:
Nvm, I see what you mean. I'll change it. Thanks a lot :)